server/session/player.go

```diff
@@ -576,15 +593,79 @@ func stackToItem(it protocol.ItemStack) item.Stack {
 	return nbtconv.ReadItem(it.NBTData, &s)
 }
 
+// itemToRecipeIngredientItem converts a recipe.InputItem into a type that can be used over the protocol.
+func itemToRecipeIngredientItem(s recipe.InputItem) protocol.RecipeIngredientItem {
+	if s.Item() == nil {
+		return protocol.RecipeIngredientItem{}
+	}
+	rid, meta, ok := world.ItemRuntimeID(s.Item())
+	if !ok {
+		panic("should never happen")
+	}
+
+	if s.Variants {
+		meta = craftingFlagAll
+	}
+
+	return protocol.RecipeIngredientItem{
+		NetworkID:     rid,
+		MetadataValue: int32(meta),
+		Count:         int32(s.Count()),
+	}
+}
+
+// itemsToRecipeIngredientItems converts a list of recipe.Items into a type that can be used over the protocol.
+func itemsToRecipeIngredientItems(s []recipe.InputItem) (r []protocol.RecipeIngredientItem) {
+	for _, st := range s {
+		r = append(r, itemToRecipeIngredientItem(st))
+	}
+	return
+}
+
+// deleteDamage strips the damage from a protocol item.
+func deleteDamage(st protocol.ItemStack) protocol.ItemStack {
+	delete(st.NBTData, "Damage")
+	return st
+}
+
+// protocolRecipes returns all recipes as protocol recipes.
+func (s *Session) protocolRecipes() []protocol.Recipe {
+	recipeList := make([]protocol.Recipe, 0, len(recipe.Recipes()))
+	for index, i := range recipe.Recipes() {
+		networkID := uint32(index) + 1
+		s.recipeMapping[networkID] = i
+
+		switch newRecipe := i.(type) {
+		case *recipe.ShapelessRecipe:
+			recipeList = append(recipeList, &protocol.ShapelessRecipe{
+				RecipeID:        uuid.New().String(),
+				Input:           itemsToRecipeIngredientItems(newRecipe.Inputs()),
+				Output:          []protocol.ItemStack{deleteDamage(stackFromItem(newRecipe.Output()))},
+				Block:           "crafting_table", // This won't always be the case in the future, but for now it's fine.
```

Comment:
Consider adding the block used to the recipes somehow, that'll save us work in the future when we do implement other crafting stations.